react/src/lib/ListItem/index.js

```diff
@@ -70,19 +70,79 @@ class ListItem extends React.Component {
     parentOnSelect && parentOnSelect(e, { value, label, eventKey });
   }
 
-  handleKeyDown = (e, eventKey) => {
+  changeTabIndex = (tabbableChildren, index) => {
+    for (let i = 0; i < tabbableChildren.length; i++) {
+      if (tabbableChildren[i].tabIndex === (index === 0 ? -1 : 0)) {
+        tabbableChildren[i].tabIndex = index;
+      }
+    }
+  }
+
+  handleKeyDown = (e, eventKey, focusLockTabbableChildren, tabbableChildrenQuery) => {
     const { disabled, onKeyDown, parentKeyDown, value, label } = this.props;
 
-    if(disabled) {
+    if (disabled) {
       e.preventDefault();
       e.stopPropagation();
     }
+    
+    if (focusLockTabbableChildren && e.target) {
+      const currListItem = e.target.closest('.md-list-item');
+
+      if  (currListItem) {
+        const tabbableChildren = currListItem.querySelectorAll(tabbableChildrenQuery);
+        if (tabbableChildren.length) {
+          if (e.keyCode === 9 && !e.shiftKey) { // TAB only
+            // only allow focus of tabbable children if TAB on the current listitem
+            if (e.target.classList.contains('md-list-item')) {
+              this.changeTabIndex(tabbableChildren, 0);
+            }
+          } else if (e.keyCode === 9 && e.shiftKey) { // SHIFT + TAB
+            // If we are on one of the tabbable children
+            if (e.target === tabbableChildren[0]) {
+              e.preventDefault();
+              e.stopPropagation();
+              // focus on the tabbable children's associated lisitem
+              e.target.closest('.md-list-item').focus();
+              // If we are on a listitem, SHIFT + TAB exits the list
+              // so we change tabindex of children to -1
+            } else if (e.target.classList.contains('md-list-item')) {
+              this.changeTabIndex(tabbableChildren, -1);
+            }
+          } else if (e.keyCode === 38 || e.keyCode == 40) {
+            // UP/DOWN guarantees change of listitem, so we change tabindex of children to -1
+            this.changeTabIndex(tabbableChildren, -1);
+          }
+        }
+      }
+    }
 
     e.persist();
     onKeyDown && onKeyDown(e);
     parentKeyDown && parentKeyDown(e, { value, label, eventKey });
   }
 
+  handleBlur = (e, focusLockTabbableChildren, tabbableChildrenQuery) => {
+    const { onBlur } = this.props;
+
+    if (focusLockTabbableChildren) {
+      if (e.target) {
+        const currListItem = e.target.closest('.md-list-item');
+        const tabbableChildren = currListItem.querySelectorAll(tabbableChildrenQuery);
```

Comment:
If someone uses list item outside of a list, I think this will throw an error since `currListItem` will be undefined

Reply:
Probably need a check for tabbableChildren too? Is it possible that someone passes the prop but not the query?

Reply:
Forgot about that adding checks now